File: alanEappen/assignments/cyclicSort/lc287/lc287.js
// Problem Statement #

// We are given an unsorted array containing ‘n+1’ numbers taken from the range 1 to ‘n’. The array has only one duplicate but it can be repeated multiple times. Find that duplicate number without using any extra space. You are, however, allowed to modify the input array.

//given: nums -> unsorted array
//goal: is to find the duplicate value 

/*
variables: curPtr = 0
while loop conditional with curPtr < nums.length
let swapPtr = nums[curPtr]-1
The if conditional (nums[curPtr] !== nums[swapPtr])
[[numscurPtr],[numswapPtr]] = [[numsswapPtr], [[numscurPtr]]]
else{
 curPtr++
}

//make another for loop to find the duplicated value 
//by going through the sorted nums
then we make a conditional if nums[i] is equal to he next element
  if(num[i] == nums[i+1])
  {
    return nums[i] 
  }
*/
//edge cases
/*
nums.length = 0 return -1
there is no duplicates return -1
*/
const find_duplicate = function (nums) {
  let curPtr = 0

  if (nums.length == 0) return -1;

  while (curPtr < nums.length) {
    let swapPtr = nums[curPtr] - 1
    if (nums[curPtr] !== nums[swapPtr]) {
      [nums[curPtr], nums[swapPtr]] = [nums[swapPtr], nums[curPtr]]
    }
    else {
      curPtr++
    }
  }

  for (let i = 0; i < nums.length; i++) {
    // console.log(i + ' = ' + nums[i])
    if (nums[i] == nums[nums.length - 1] && i < nums.length - 1) {
      return nums[i]
    }
  }

  return -1;
};

console.log(find_duplicate([1, 4, 4, 3, 2])); //4    
console.log(find_duplicate([2, 1, 3, 3, 5, 4])); //3
console.log(find_duplicate([2, 4, 1, 4, 4])); //4
console.log(find_duplicate([2, 4, 1, 3, 5])); //-1
console.log(find_duplicate([])); //-1


// Solution
// -----
// function find_duplicate(nums) {
//   let i = 0;
//   while (i < nums.length) {
//     if (nums[i] !== i + 1) {
//       j = nums[i] - 1;
//       if (nums[i] !== nums[j]) {
//         [nums[i], nums[j]] = [nums[j], nums[i]]; // swap
//       } else { // we have found the duplicate
//         return nums[i];
//       }
//     } else {
//       i += 1;
//     }
//   }
//   return -1;
// }

// -----

// Time complexity #
// The time complexity of the above algorithm is O(n).

// Space complexity #
// The algorithm runs in constant space O(1) but modifies the input array.

File: taojieTan/assignments/cyclicSort/lc287/lc287.py
# Problem Statement #

# We are given an unsorted array containing ‘n+1’ numbers taken from the range 1 to ‘n’. The array has only one duplicate but it can be repeated multiple times. Find that duplicate number without using any extra space. You are, however, allowed to modify the input array.

# input: an unsorted list with integers ranging from 1 to n and length of n+1
# output: the duplicate integer in the list
# edge cases: 1. length of list is smaller than 2 => return 0

'''
 Since we are allowed to modify the list, we can first place each integer into correct position (integer 1 would be on index 0, 2 would be on index 1, and so on),
 and if we have another integer that needs to be swapped to the same position, it means we have duplicate integer. 

 We can use a while loop to traverse through the list, and check if each integer is in correct position, 
 if not, we swap it with the index equals to integer - 1, 
 and the meantime, we also need to check if the value in the index (integer - 1) has the same value as integer, 
 if it does, we find the duplicate and just return the integer; Else, we can swap them.
 If the integer already in the correct position, we increment the index to keep traversing the list. 

 we will return 0 if no duplicate found after the loop.
'''

def find_duplicate(nums):
  # TODO: Write your code here
  index = 0

  while index < len(nums):
    if nums[index] != index + 1:
      correct_pos = nums[index] - 1
      if nums[correct_pos] == nums[index]:
        return nums[index]
      else:
        nums[correct_pos], nums[index] = nums[index], nums[correct_pos]
    else:
      index += 1

  return -1

def main():
  print(find_duplicate([1, 4, 4, 3, 2]))
  print(find_duplicate([2, 1, 3, 3, 5, 4]))
  print(find_duplicate([2, 4, 1, 4, 4]))
  print(find_duplicate([2, 1, 3, 1, 5, 4]))
  print(find_duplicate([4, 1, 1, 1, 1]))
  print(find_duplicate([2, 1]))
  print(find_duplicate([1]))
  print(find_duplicate([]))


main()


# Solution
# -----

# def find_duplicate(nums):
#   i = 0
#   while i < len(nums):
#     if nums[i] != i + 1:
#       j = nums[i] - 1
#       if nums[i] != nums[j]:
#         nums[i], nums[j] = nums[j], nums[i]  # swap
#       else:  # we have found the duplicate
#         return nums[i]
#     else:
#       i += 1

#   return -1

# -----

# Time complexity #
# The time complexity of the above algorithm is O(n).

# Space complexity #
# The algorithm runs in constant space O(1) but modifies the input array.
